lnwallet/channel.go

```diff
@@ -2151,22 +2170,74 @@ func (lc *LightningChannel) closeObserver(channelCloseNtfn *chainntnfs.SpendEven
 			"broadcast!!!", lc.channelState.FundingOutpoint,
 			remoteStateNum)
 
+		if err := lc.channelState.MarkBorked(true); err != nil {
+			lc.Unlock()
+			walletLog.Errorf("unable to mark channel as borked: %v", err)
+			return
+		}
+
+		settledBalance := lc.channelState.LocalCommitment.LocalBalance.ToSatoshis()
+		closeSummary := channeldb.ChannelCloseSummary{
+			ChanPoint:      lc.channelState.FundingOutpoint,
+			ChainHash:      lc.channelState.ChainHash,
+			ClosingTXID:    *commitSpend.SpenderTxHash,
+			RemotePub:      lc.channelState.IdentityPub,
+			Capacity:       lc.Capacity,
+			SettledBalance: settledBalance,
+			CloseType:      channeldb.BreachClose,
+			IsPending:      true,
+		}
+
 		// Create a new reach retribution struct which contains all the
 		// data needed to swiftly bring the cheating peer to justice.
 		retribution, err := newBreachRetribution(lc.channelState,
 			broadcastStateNum, commitTxBroadcast)
 		if err != nil {
+			lc.Unlock()
 			walletLog.Errorf("unable to create breach retribution: %v", err)
 			return
 		}
+		lc.Unlock()
 
 		walletLog.Debugf("Punishment breach retribution created: %v",
 			spew.Sdump(retribution))
 
 		// Finally, send the retribution struct over the contract beach
 		// channel to allow the observer the use the breach retribution
 		// to sweep ALL funds.
-		lc.ContractBreach <- retribution
+		select {
+		case lc.ContractBreach <- retribution:
+		case <-lc.quit:
+			walletLog.Errorf("channel shutting down")
+			return
+		}
+
+		select {
+		case err, ok := <-retribution.Err:
+			if ok {
```

Comment:
better to check `err != nil`? Not sure what is most common

Reply:
Could probably add a comment, but this only ever sends back non-nil errors across the channel, and closes them upon success. This makes successful handoffs to asynchronously ACK back, without having to send nil.